### DISCUSSION TITLE
MPI allreduce cost function

### DISCUSSION THREAD

**Discussion by sss441803** [NVIDIA/cuQuantum#78] [Q&A] [answered]
Hi,

I ran some tests and the contraction time of path order optimized with the cost function time heuristic is about 5 times faster than optimized with flops. However, when performing `allreduce`, I can only find `opt_cost` attribute of the optimization results of each rank. Are there ways to find the optimal time contraction path with MPI?

Sincerely,
Henry

**Comment by haidarazzam** [comment]
Dear Minzhao,
Can you please elaborate in more detail what you need?
`Are there ways to find the optimal time contraction path with MPI?`
it should be possible, even when cost function equal time, we still return the estimated runtime or the effective flops which are correlated.
Thanks

**Reply by sss441803** [reply]
Hi Azzam,

Thanks for the reply! I just want the result from the rank with the shortest run time. I am just not sure if path that has the lowest info.opt_cost is the best run time one as well. Right now I'm keeping the result from the MPI rank with the lowest opt_cost, which I thought is flops. Is it effective flops?

**Reply by sss441803** [reply]
If opt_cost is effective flops, then it's not real flops right? How would I calculate the flops efficiency in that case? Thank you in advance for your help!

**Reply by haidarazzam** [reply, marked as answer]
Minzhao,
Currently we allreduce'd for conventional flops, thus it might not be the optimal when you use cost=time.
This will be fixed in the next patch/release.
However the workaround it now is to performs the reduce yourself and feed it to the optimizerInfo.
We return both effective flops and conventional flops.
Effective flops is:

- same as conventional when cost=flops
- reflecting estimated runtime when cost=time (effective=runtime_estimation*peak_of_machine) so it is always going to looks like larger than conventional_flops, but it is a consistent value that we can use to compare different path and different runs in a consistent way. effective_flops=  runtime_estimation x constant
Azzam

**Reply by mtjrider** [reply]
@sss441803 did @haidarazzam answer your question satisfactorily?
If so, please mark his response as the answer.
If not, can you provide additional clarity or pose any lingering/relevant questions?

**Comment by sss441803** [comment]
Hi! Just a follow-up as I am trying to calculate the numerical efficiency. I have A100. What is the theoretical peak FLOPS? Is it FP64 in the datasheet which is 9.7 TFLOPS?

When tensors are complex64 and complex128, the resulting opt_cost is the same, but actual contraction is obviously faster for complex64. If I finish contraction of complex128 in t seconds, should the efficiency be info.opt_cost / t / 9.7 TFLOPS? And for complex64, info.opt_cost / t 19.5 TFLOPS?

**Reply by haidarazzam** [reply]
the cost (flops) as returned by cutensorNet is the number of floating point operations thus it is the same whether you are doing single or double precision.
Note that as mentioned in the doc, we return the cost for real operation (not complex), thus to compute the efficiency of complex operation, you needs to convert the cost into complex cost (multiply flops by 4) then you can compute the efficiency.
Thus for complex64 it should be (4*info.opt_cost) / t / 19.5 TFLOPS
